File: back/src/common/services/geocoding.service.ts
import { Injectable, Logger, BadRequestException } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';

export interface GeocodeResult {
  latitude: number;
  longitude: number;
  city: string;
  country: string;
  displayName: string;
}

@Injectable()
export class GeocodingService {
  private readonly logger = new Logger(GeocodingService.name);

  constructor(private configService: ConfigService) {}

  /**
   * Convert city name to coordinates using OpenStreetMap Nominatim API
   * This is a free geocoding service that doesn't require an API key
   */
  async geocodeCity(cityName: string): Promise<GeocodeResult> {
    if (!cityName || cityName.trim().length === 0) {
      throw new BadRequestException('City name is required');
    }

    const cleanCityName = cityName.trim();
    
    try {
      // Use Nominatim API (OpenStreetMap's geocoding service)
      const url = `https://nominatim.openstreetmap.org/search?format=json&q=${encodeURIComponent(cleanCityName)}&limit=1&addressdetails=1`;
      
      const response = await fetch(url, {
        headers: {
          'User-Agent': 'MusicRoom/1.0 (https://github.com/smiley16479/music-room)', // Required by Nominatim
        },
      });

      if (!response.ok) {
        throw new Error(`Geocoding API returned status ${response.status}`);
      }

      const data = await response.json() as any[];

      if (!data || data.length === 0) {
        throw new BadRequestException(`City "${cleanCityName}" not found. Please check the spelling or try a more specific name.`);
      }

      const result = data[0];
      
      // Validate the result has coordinates
      if (!result.lat || !result.lon) {
        throw new BadRequestException(`Unable to get coordinates for "${cleanCityName}"`);
      }

      const latitude = parseFloat(result.lat);
      const longitude = parseFloat(result.lon);

      // Validate coordinates are within valid ranges
      if (latitude < -90 || latitude > 90 || longitude < -180 || longitude > 180) {
        throw new BadRequestException(`Invalid coordinates received for "${cleanCityName}"`);
      }

      // Extract city and country information
      const address = result.address || {};
      const city = address.city || address.town || address.village || address.municipality || cleanCityName;
      const country = address.country || 'Unknown';
      const displayName = result.display_name || `${city}, ${country}`;

      this.logger.log(`Successfully geocoded "${cleanCityName}" to ${latitude}, ${longitude}`);

      return {
        latitude,
        longitude,
        city,
        country,
        displayName,
      };
    } catch (error) {
      this.logger.error(`Failed to geocode city "${cleanCityName}": ${error.message}`);
      
      if (error instanceof BadRequestException) {
        throw error;
      }
      
      throw new BadRequestException(`Unable to find location for "${cleanCityName}". Please try a different city name.`);
    }
  }

  /**
   * Reverse geocoding: convert coordinates to city name
   * Useful for displaying readable location names
   */
  async reverseGeocode(latitude: number, longitude: number): Promise<GeocodeResult> {
    if (latitude < -90 || latitude > 90 || longitude < -180 || longitude > 180) {
      throw new BadRequestException('Invalid coordinates provided');
    }

    try {
      const url = `https://nominatim.openstreetmap.org/reverse?format=json&lat=${latitude}&lon=${longitude}&addressdetails=1`;
      
      const response = await fetch(url, {
        headers: {
          'User-Agent': 'MusicRoom/1.0 (https://github.com/smiley16479/music-room)',
        },
      });

      if (!response.ok) {
        throw new Error(`Reverse geocoding API returned status ${response.status}`);
      }

      const result = await response.json() as any;

      if (!result || !result.address) {
        throw new BadRequestException('Unable to determine location from coordinates');
      }

      const address = result.address;
      const city = address.city || address.town || address.village || address.municipality || 'Unknown Location';
      const country = address.country || 'Unknown';
      const displayName = result.display_name || `${city}, ${country}`;

      return {
        latitude,
        longitude,
        city,
        country,
        displayName,
      };
    } catch (error) {
      this.logger.error(`Failed to reverse geocode coordinates ${latitude}, ${longitude}: ${error.message}`);
      
      if (error instanceof BadRequestException) {
        throw error;
      }
      
      throw new BadRequestException('Unable to determine location from coordinates');
    }
  }

  /**
   * Validate that a city name can be geocoded
   * Useful for form validation without storing the result
   */
  async validateCity(cityName: string): Promise<boolean> {
    try {
      await this.geocodeCity(cityName);
      return true;
    } catch (error) {
      return false;
    }
  }
}

File: flutter_app_/lib/core/services/geocoding_service.dart
import 'package:flutter/foundation.dart';
import 'package:geocoding/geocoding.dart';
import 'package:http/http.dart' as http;
import 'dart:convert';

/// Geocoding Service - converts place names to coordinates
class GeocodingService {
  /// Convert a place name (city, address) to latitude/longitude
  /// Returns [latitude, longitude] or null if not found
  Future<Map<String, double>?> getCoordinatesFromPlace(String placeName) async {
    try {
      debugPrint('🔍 Geocoding place: $placeName');

      if (kIsWeb) {
        return await getCoordinatesFromPlaceForWeb(placeName);
      }

      if (placeName.isEmpty) {
        debugPrint('⚠️ Empty place name');
        return null;
      }

      // locationFromAddress returns a list of Placemarks
      final locations = await locationFromAddress(placeName);

      debugPrint('📍 Locations found: ${locations.length}');

      if (locations.isEmpty) {
        debugPrint('⚠️ No location found for: $placeName');
        return null;
      }

      final location = locations.first;
      final lat = location.latitude;
      final lon = location.longitude;

      debugPrint('✅ Found location: $lat, $lon');

      return {'latitude': lat, 'longitude': lon};
    } catch (e) {
      debugPrint('❌ Geocoding error: $e (${e.runtimeType})');
      return null;
    }
  }

  /// Convert a place name to coordinates using Nominatim (OpenStreetMap)
  Future<Map<String, double>?> getCoordinatesFromPlaceForWeb(
    String placeName,
  ) async {
    if (placeName.trim().isEmpty) return null;

    try {
      debugPrint('🌐 Web geocoding: $placeName');

      final uri = Uri.parse(
        'https://nominatim.openstreetmap.org/search'
        '?q=${Uri.encodeComponent(placeName)}'
        '&format=json'
        '&limit=1',
      );

      final response = await http.get(
        uri,
        headers: {
          // Obligatoire selon la policy Nominatim
          'User-Agent': 'music-room-app/1.0 (contact@musicroom.app)',
        },
      );

      if (response.statusCode != 200) {
        debugPrint('❌ HTTP error: ${response.statusCode}');
        return null;
      }

      final List data = jsonDecode(response.body);

      if (data.isEmpty) {
        debugPrint('⚠️ No results for $placeName');
        return null;
      }

      final result = data.first;

      return {
        'latitude': double.parse(result['lat']),
        'longitude': double.parse(result['lon']),
      };
    } catch (e) {
      debugPrint('❌ Web geocoding error: $e');
      return null;
    }
  }

  /// Get place name from coordinates (reverse geocoding)
  /// Returns place name or null if not found
  Future<String?> getPlaceFromCoordinates({
    required double latitude,
    required double longitude,
  }) async {
    try {
      debugPrint('🔍 Reverse geocoding: $latitude, $longitude');

      // placemarkFromCoordinates returns a list of Placemarks
      final placemarks = await placemarkFromCoordinates(latitude, longitude);

      debugPrint('📍 Placemarks found: ${placemarks.length}');

      if (placemarks.isEmpty) {
        debugPrint('⚠️ No place found for coordinates');
        return null;
      }

      final placemark = placemarks.first;
      final place =
          placemark.locality ??
          placemark.administrativeArea ??
          placemark.country ??
          'Unknown';
      debugPrint('✅ Found place: $place');

      return place;
    } catch (e) {
      debugPrint('❌ Reverse geocoding error: $e (${e.runtimeType})');
      return null;
    }
  }
}
